java-checks/src/main/java/org/sonar/java/checks/SynchronizedClassUsageCheck.java

```diff
@@ -57,15 +56,20 @@
 
   private final Deque<Set<String>> exclusions = new ArrayDeque<>();
 
-  private Set<Tree> visited = new HashSet<>();
+  private final Set<Tree> visited = new HashSet<>();
 
   @Override
   public List<Tree.Kind> nodesToVisit() {
-    return Arrays.asList(Tree.Kind.CLASS, Tree.Kind.ENUM, Tree.Kind.INTERFACE, Tree.Kind.ANNOTATION_TYPE);
+    // We register on compilation units to clear the visited data structure when scanning a new file
+    return Arrays.asList(Tree.Kind.CLASS, Tree.Kind.COMPILATION_UNIT, Tree.Kind.ENUM, Tree.Kind.INTERFACE, Tree.Kind.ANNOTATION_TYPE);
   }
 
   @Override
   public void visitNode(Tree tree) {
+    if (tree.is(Tree.Kind.COMPILATION_UNIT)) {
+      // We clear the visited data structure when entering every file
+      visited.clear();
+    }
```

Comment:
No! You should return!
If not, you are going to trigger the `exclusionsVisitor` on a way too large scope, and end up scanning everything!
You can also probably get rid of the comment, the body of the if is self-explanatory.

```java
   if (tree.is(Tree.Kind.COMPILATION_UNIT)) {
      visited.clear();
      return;
   }